python/pyarrow/dataset/protocol.py

```diff
@@ -0,0 +1,77 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Protocol definitions for pyarrow.dataset
+
+These provide the abstract interface for a dataset. Other libraries may implement
+this interface to expose their data, without having to extend PyArrow's classes.
+
+Applications and libraries that want to consume datasets should accept datasets
+that implement these protocols, rather than requiring the specific
+PyArrow classes.
+"""
+from abc import abstractmethod
+from typing import Iterator, List, Optional, Protocol
+
+from pyarrow.dataset import Expression
+from pyarrow import Table, IntegerArray, RecordBatch, RecordBatchReader, Schema
+
+
+class Scanner(Protocol):
+    @abstractmethod
+    def count_rows(self) -> int:
+        ...
+    
+    @abstractmethod
+    def head(self, num_rows: int) -> Table:
+        ...
+
+    @abstractmethod
+    def take(self, indices: IntegerArray) -> Table:
+        ...
+    
+    @abstractmethod
+    def to_table(self) -> Table:
+        ...
+    
+    @abstractmethod
+    def to_batches(self) -> Iterator[RecordBatch]:
+        ...
+
+    @abstractmethod
+    def to_reader(self) -> RecordBatchReader:
+        ...
```

Comment:
TODO: do we need a protocol for RecordBatchReader to make sure that implementors can provide a RecordBatchReader for whom we can call `_export_to_c` and get back an implementation that isn't constrained by the GIL?

Reply:
Yes. My question here is if you do:

```python
reader = RecordBatchReader._import_from_c(ptr)
new_ptr = 0
reader._export_from_c(new_ptr)
```

Is `new_ptr` just the a pointer to the original stream? Or did it get wrapped in some Python RBR that will lock the GIL when it's methods are called?

Reply:
I suspect that will depend on the underlying record batch reader.  Is there any reason the record batch reader can't hold the GIL?  What if it is a pure-python implementation?

Reply:
Looking through the code, it seems like it will be the original reader wrapped in some C++ class, but not at all wrapped in Python which is good. I think it's fine if there is Python, but wanted to make sure those that are implemented in Rust or C++ don't have such a limitation.